### DISCUSSION TITLE
how to use a JavaScript module built for node.js

### DISCUSSION THREAD

**Discussion by barbalex** [tauri-apps/tauri#2592] [Q&A] [answered]
My app now runs on electron. I am considering using tauri instead.

The app uses https://github.com/JoshuaWise/better-sqlite3 which is meant to run in node.

Is there a way I can use it with tauri?

**Comment by Shotman** [comment, marked as answer]
Tauri doesn't use nodeJS for server stuff, in fact Tauri doesn't use a localhost server, you have to use Rust crates and redo your SQLite logic with the rust backend

**Reply by barbalex** [reply]
better-sqlite3 is a library, not a server. The logic is inside the (client side) JavaScript code of my app, not in any node server.
The problem is that better-sqlite3 can not be used in a browser. Without node available, it will fail to run.

**Reply by Shotman** [reply]
That's what I mean by node server, tauri doesn't use node binary at all, so either you code your app in node package it and use the sidecar feature or you use Rust

**Reply by barbalex** [reply]
> either you code your app in node package it and use the sidecar feature 

Which would not work as it also needs browser methods.

> or you use Rust

Which would mean a total rewrite. Then I can just as well use something like flutter.

I am going to remain on electron in that case.

But thank you a lot for your great help!

**Reply by barbalex** [reply]
By the way: It might be helpful for other electron users considering to migrate to tauri to point out the fact that tauri is not a valid choice if your app uses JavaScript libraries built for node.js (as is possible in electron).

**Reply by adityatoshniwal** [reply]
Even I installed tauri to my existing project to try out, but found this. I have my backend services in node/express and I replace the electron node server with my own. Since it is not possible with tauri, need to stick with electron then.
I agree. This should be mentioned on the website.

**Reply by igaul** [reply]
Link to where its mentioned with an example: https://tauri.studio/en/docs/guides/bundler/sidecar/

**Reply by CMCDragonkai** [reply]
@igaul that link no longer works. Where is this information?

**Reply by igaul** [reply]
https://tauri.studio/guides/building/sidecar/

**Reply by xeoshow** [reply]
> https://tauri.studio/guides/building/sidecar/

@igaul Thanks for the information. By Using Node.js on a Sidecar, is there any negative effect for such solution? My project uses html,js,css and communicate with nodejs backend via restful api.

Great thanks!

**Reply by igaul** [reply]
@xeoshow just package size, as you'll need to bundle a node runtime via something like [pkg](https://github.com/vercel/pkg) or use [deno](https://github.com/denoland/deno) if that's an option.

If your backend server is not local, then you can use fetch api like normal or use tauri messages and a simple rust web framework like [axum](https://github.com/tokio-rs/axum) to either handle or proxy your requests ( and save something like 50mb ).

**Reply by marcincodes** [reply]
I stumbled upon this issue when I was searching for a similar problem so I made a project called Taurine: https://github.com/Perfect7M/taurine
It's setup by packing your code and NodeJS server into a single executable like described above but in a more automatic way. Read more here about Taurine: https://marcin.codes/posts/tauri-+-nodejs:-alternative-to-electron/

**Comment by xeoshow** [comment]
@igaul Thanks so much for reply. My desktop GUI application previously used electron, and now migrating to tauri with nodejs locally. So seems what you provided is a feasible solution? Regarding the sidecar solution, is there anyway that we could start the nodejs when initially starting the application? Also I will need the node modules work normally for them, is this possible?

Best regards, Jason

**Reply by igaul** [reply]
@xeoshow yeah, the tauri team has a few [examples](https://github.com/tauri-apps/tauri/blob/dev/examples/sidecar/src-tauri/src/main.rs) floating around, but essentially you'll want to bundle your node app with all dependencies ( I use [esbuild](https://github.com/evanw/esbuild) ) -- this should be about the same as you do for your electron preload file -- and use pkg to bundle it together with runtime... you then have a single file that you spawn in tauri main.
